File: algorithms/Python/sorting/insertion_sort.py
"""
Analogizing this algorithm with inserting a playing 
card into your hand, we distinguish the "key" as
the inserting card and find the position of that 
card among the previous j - 1 cards. 

O(n^2) runtime (the deck is sorted in descending order).
"""
arr = [12, 3, 7, 22, -12, 100, 1]


def insertion_sort(arr):
	"""
	>>> insertion_sort(arr)
	>>> arr
	[-12, 1, 3, 7, 12, 22, 100]
	"""
	for j in range(1, len(arr)):
		key = arr[j]
		# insert the key into the sorted sequence arr[1, ... , j - 1]
		i = j - 1
		while i >= 0 and arr[i] > key:
			arr[i + 1] = arr[i]
			i -= 1
		arr[i + 1] = key


if __name__ == "__main__":
	insertion_sort(arr)
	print("Sorted array: ")
	for ele in arr:
		print(f"\t{ele}")

File: algorithms/Go/sorting/insertion-sort.go
/*
* Analogizing this algorithm with inserting a playing
* card into your hand, we distinguish the "key" as
* the inserting card and find the position of that
* card among the previous j - 1 cards.

* O(n^2) runtime (the deck is sorted in descending order).
 */

package sorting

import (
	"fmt"
)

func insertionSort(arr []int) {
	n := len(arr)
	for i := 1; i < n; i++ {
		key := arr[i]
		j := i - 1
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
}

func RunInsertionSort() {
	arr := []int{10, 1, 6, 256, 2, 53, 235, 53, 1, 7, 23}
	fmt.Println("Unsorted Array:", arr)
	insertionSort(arr)
	fmt.Println("Sorted Array:", arr)
}
